### PR TITLE
[ADD] edi_stock: Move report document and Move cancellation report

### PR DESCRIPTION
**Pull request by PeterAlabaster**
The concept currently exists where EDI can track individual picks which
have been reported on with the `_edi_pick_report_via`. This extends that
concept with a new model `edi.move.report.document` which makes use of
`_edi_move_report_via`.
The same mechanism is used for this, however instead
of having a pick --> moves, and binding the edi doc id to the pick, we
simply have moves, and bind the edi doc id to the moves.

Added ontop of this concept is a move cancellation report, which extends
the base class slightly to generate an outbound email for any new move
cancellations against picking types which have the appropriate config.
The email template which is used is also configured on the picking type,
so it is possible to create different email templates for different pick
types, and have them all batched and sent by the same edi document.

A default mail template is included too, which has some sane defaults.

Signed-off-by: Peter Alabaster <peter.alabaster@unipart.io>

User-story: 14633

### CODE REVIEW

File: addons/edi_stock/data/edi_move_cancellation_data.xml [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0"?>
+<odoo>
+    <data>
+        <!-- Processed Purchase Orders sequence -->
```
Comment (PeterAlabaster) [inline addons/edi_stock/data/edi_move_cancellation_data.xml]: oops.. will fix these

File: addons/edi_stock/data/edi_move_cancellation_data.xml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            <field name="name">Move Cancellation</field>
+            <field name="model_id" ref="model_edi_record_stock_cancellation" />
+            <field name="auto_delete" eval="False" />
+            <field name="email_from">${user.name} &lt;${user.email | safe}&gt;</field>
+            <field name="subject">Move Cancellation for ${object.sale_id.client_order_ref | safe}</field>
+            <field name="body_html">
+                <![CDATA[
+        <p>Products have been cancelled from a move related to <span style="font-weight: bold;">${object.sale_id.client_order_ref}</span> for the following customer:</p>
+        <p>${object.customer_id.name}<br/>
+            % if object.customer_id.street:
```
Comment (PeterAlabaster) [inline addons/edi_stock/data/edi_move_cancellation_data.xml]: i couldn't find any nice examples of address formatting in jinja, so opted to do this. if anyone knows better ways, let me know!

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        StockPickingType = self.env["stock.picking.type"]
+        picking_types_to_consider = StockPickingType.search(
+            [("u_enable_pick_cancellation_email_notifs", "=", True)]
+        )
+        domain = [
+            ("state", "=", "cancel"),
+            ("picking_type_id", "in", picking_types_to_consider.ids),
+        ]
+        if self._edi_move_report_via is not None:
```
Comment (PeterAlabaster) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: just realised i do not need this, will change and just add it directly to the domain

File: addons/edi_stock/__manifest__.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # pylint: disable=missing-docstring,pointless-statement
 {
-    'name': "EDI for Inventory Management",
```
Comment (PeterAlabaster) [inline addons/edi_stock/__manifest__.py]: Note to self: Split formatting into its own commit

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    )
+
+
+class EdiStockCancellationDocument(models.AbstractModel):
+    """Stock Cancellation document model
+
+    Stock Cancellation documents contain sets of moves
+    (grouped by sale & picking_type) which have been cancelled
+    since the previous EDI export. Note this report only makes sense
+    for outgoing picks linked to sales
```
Comment (PeterAlabaster) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: Note to self, this comment needs updating to reflect the fact this report now works for incoming/outgoing picks unrelated to sales

File: addons/edi_stock/data/edi_move_cancellation_data.xml (hunk 2 above)
Comment (PeterC10) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: I don't think there's a nice way to do many things in jinja tbh. :smile: 

Maybe long term we could look into adding a computed field, something like `formatted_address` which would take care of this for us and just let us use one line.

File: addons/edi_stock/models/edi_move_report_document.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    @api.model
+    def prepare(self, doc):
+        """Prepare document"""
+        MoveReport = self.move_report_record_model(doc)
+        Move = self.env["stock.move"]
+        # Lock moves to prevent concurrent report generation attempts
+        moves = Move.search(self.move_report_domain(doc), order="id")
+        if self._edi_move_report_via is not None:
+            moves.write({self._edi_move_report_via: False})
+        movelist = (x.with_prefetch(moves._prefetch) for x in self.move_report_list(doc, moves))
```
Comment (PeterC10) [inline addons/edi_stock/models/edi_move_report_document.py]: Suggestion: Use `mr` instead of `x` as we're dealing with move reports.

**Review by PeterC10 — APPROVED** [review]

File: addons/edi_stock/data/edi_move_cancellation_data.xml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+          ref('edi_record_type_move_cancellation'),
+        ])]" />
+            <!-- pick_type_ids will be determined based on which ones have
+            u_enable_pick_cancellation_email_notifs set -->
+        </record>
+    </data>
+
+    <data noupdate="1">
+        <!-- Intended as a baseline template for Move Cancellation EDI, in a noupdate
+        block so that recipients can be configured without being overridden on upgrade -->
```
Comment (adamrpatrick) [inline addons/edi_stock/data/edi_move_cancellation_data.xml]: Just check here that we can still set some of these in a customer module (if that's intended). Odoo can be funny with noupdates, sometimes setting it in one place work across all places we configure the specific record.

**Review by BlommaertsEdwin — APPROVED** [review]

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+class EdiDocument(models.Model):
+    """Extend ``edi.document`` to include stock move cancellation records"""
+
+    _inherit = "edi.document"
+
+    move_cancellation_ids = fields.One2many(
+        "edi.record.stock.cancellation",
+        "doc_id",
+        string="Cancelled Stock Moves",
+    )
```
Comment (adamrpatrick) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: Are these actually the ids of cancelled stock moves? If not the maybe describe this in the string, or if they are rename the field to something like `cancelled_move_ids`.

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+    @api.model
+    def move_report_list(self, doc, moves):
+        """
+        Inherit to group moves by sale & picking_type to result
+        in a cancellation record for each unique combination
+        """
+        res = super().move_report_list(doc, moves)
+        grouped_moves = []
+        for sales_order, moves_by_sale in res.groupby(lambda move: move.sale_line_id.order_id):
```
Comment (adamrpatrick) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: Not all customers that have `edi_stock` have sale orders. Would this still work for those customers, or does this need moving into its own module/ a different module?

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+class EdiStockCancellationRecord(models.Model):
+    """Stock Cancellation record"""
+
+    _name = "edi.record.stock.cancellation"
+    _inherit = "edi.move.report.record"
+    _description = "Stock Cancellation"
+
+    order_reference = fields.Char(
+        string="Order Reference", required=True, readonly=True, index=True
```
Comment (adamrpatrick) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: General we overuse indexes, we don't need them on _every_ field. It's something that needs looking at another time though and not here... just wanted to moan...

File: addons/edi_stock/data/edi_move_cancellation_data.xml (hunk 7 above)
Comment (PeterAlabaster) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: Yeah, usually I would just remove the noupdate and add an xml inherit to update it for a customer.. However - for this specific story it's been requested that:
```There's an amendable list for email addresses to receive this update```
I'll double check how it behaves without a noupdate and with manually added recipients, as because we don't specify the field here it may persist

File: addons/edi_stock/models/edi_move_cancellation_report_document.py (hunk 8 above)
Comment (PeterAlabaster) [reply addons/edi_stock/models/edi_move_cancellation_report_document.py]: Good spot, will get this renamed

File: addons/edi_stock/models/edi_move_cancellation_report_document.py (hunk 9 above)
Comment (PeterAlabaster) [reply addons/edi_stock/models/edi_move_cancellation_report_document.py]: Ah, I was thinking this in the morning but didn't think about the `sale_line_id` column not existing in those instances.

Ideally I don't want to have a bridge module for this, so might just add a check for the column existing - if that sounds OK?

File: addons/edi_stock/data/edi_move_cancellation_data.xml (hunk 2 above)
Comment (PeterAlabaster) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: I will look into `contact_address` on res.partner.

File: addons/edi_stock/models/edi_move_report_document.py (hunk 6 above)
Comment (PeterAlabaster) [reply addons/edi_stock/models/edi_move_report_document.py]: gonna go with `move` as `self.move_report_list()` returns `stock.move` iterables

File: addons/edi_stock/data/edi_move_cancellation_data.xml (hunk 7 above)
Comment (PeterAlabaster) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: tested and works fine without noupdate="1", so i'll remove

File: addons/edi_stock/data/edi_move_cancellation_data.xml (hunk 2 above)
Comment (PeterAlabaster) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: so `contact_address` works mostly, but it uses `\n` rather than `<br/>`. I am pretty (90%) sure that  `\n` isn't a valid newline for emails
I tried ```${object.customer_id.contact_address.replace("\n", "<br/>")}``` but this doesn't parse the line breaks
image

When hardcoding them in the if statements they get parsed as line breaks properly, it's probably escaping them to a string or something and preventing them from actually producing dom output..

TIL: `| safe` prevents odoo escaping db values.. So with it users can screw up the DOM by putting jinja code in their address (shrug)
`<p>${object.customer_id.contact_address.replace("\n", "<br/>") | safe}</p>` gives me the result i want, at the expense of jinja injection :-1:

Comment (PeterC10) [reply addons/edi_stock/data/edi_move_cancellation_data.xml]: That's definitely annoying! But thanks for looking into that. :)

Maybe we can just raise an internal issue, it's not a blocker and certainly shouldn't hold up this story. :)

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 11]
```diff
@@ -7,6 +7,18 @@
...
+class IrModuleModule(models.Model):
+    _inherit = "ir.module.module"
+
+    def _is_sale_stock_installed(self):
+        """
+        Returns bool() sale_stock module existing and being in state 'installed'.
+        With more time this be separated out into a bridge module
+        """
+        sale_stock_module = self.search([("name", "=", "sale_stock")])
+        return bool(sale_stock_module and sale_stock_module.state == "installed")
```
Comment (PeterC10) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: (Moot if we decide we should be adding a EDI Sale/Stock linking module)

Nit: You could just include the state in the `sale_stock_module` domain. Probably a tiny bit more efficient then doing a search and then accessing the value from the module record.

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    def move_report_list(self, doc, moves):
+        """
+        Inherit to group moves by sale & picking_type to result
+        in a cancellation record for each unique combination
+        """
+        IrModuleModule = self.env["ir.module.module"]
+        res = super().move_report_list(doc, moves)
+        grouped_moves = []
+
+        if IrModuleModule._is_sale_stock_installed():
```
Comment (ws-mps) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: It would be better to call another function to handle all of this. Preferably only handling the case without sale orders. Then create a new module edi_sale_stock where we override/extend such function to handle the sale order case.

File: addons/edi_stock/models/edi_move_cancellation_report_document.py (hunk 11 above)
Comment (ws-mps) [reply addons/edi_stock/models/edi_move_cancellation_report_document.py]: I think we should create the module.

File: addons/edi_stock/models/stock_picking.py [hunk 13]
```diff
@@ -6,39 +6,51 @@
...
-        'edi.document.type',
+        "edi.document.type",
         string="EDI document types",
     )
     edi_pick_report_autoemit = fields.Boolean(
         string="EDI autoemit",
         help="Create EDI pick report documents automatically",
         default=False,
     )
+    u_enable_pick_cancellation_email_notifs = fields.Boolean(
```
Comment (ws-mps) [inline addons/edi_stock/models/stock_picking.py]: This field should be named `x_` not `u_`.

File: addons/edi_stock/models/stock_picking.py [hunk 14]
```diff
@@ -6,39 +6,51 @@
...
+        string="Enable Pick Cancellation Notifications",
+        default=False,
+        help="""
+        When a pick using this picking type is cancelled,
+        send an email out using the template defined in u_pick_cancellation_email_notif_template
+        """,
+    )
+    u_pick_cancellation_email_notif_template_id = fields.Many2one(
```
Comment (ws-mps) [inline addons/edi_stock/models/stock_picking.py]: This field should be named `x_` not `u_`.

File: addons/edi_stock/models/edi_move_cancellation_report_document.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        }
+        if sale:
+            values.update(
+                {
+                    "order_reference": sale.client_order_ref or sale.name,
+                    "customer_id": sale.partner_id.id,
+                }
+            )
+        else:
+            move_origins = moves.filtered(lambda move: move.picking_id.origin != False).mapped(
```
Comment (ws-mps) [inline addons/edi_stock/models/edi_move_cancellation_report_document.py]: nit: != False not really needed, since fields default to False when empty.

File: addons/edi_stock/models/edi_move_cancellation_report_document.py (hunk 9 above)
Comment (PeterAlabaster) [reply addons/edi_stock/models/edi_move_cancellation_report_document.py]: discussed w/ miquel, going to add `edi_sale_stock` module in place of what i did with the nasty module check

File: addons/edi_stock/models/edi_move_cancellation_report_document.py (hunk 11 above)
Comment (PeterC10) [reply addons/edi_stock/models/edi_move_cancellation_report_document.py]: Yep can see that decision has now been made, will resolve. :)

**Review by ws-mps — APPROVED** [review]

**Review by adamrpatrick — APPROVED** [review]
